// @willsoto/nestjs-prometheus/dist/metrics/summary.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeSummaryProvider = makeSummaryProvider;
const constants_1 = require("../constants");
const utils_1 = require("./utils");
/**
 * @public
 */
function makeSummaryProvider(options) {
    const { inject: extraInject = [], collect, ...restOptions } = options;
    return {
        provide: (0, utils_1.getToken)(restOptions.name),
        useFactory(config, ...injectedDeps) {
            const metricOptions = collect
                ? {
                    ...restOptions,
                    collect() {
                        return collect.call(this, ...injectedDeps);
                    },
                }
                : restOptions;
            return (0, utils_1.getOrCreateMetric)("Summary", metricOptions, config);
        },
        inject: [
            {
                token: constants_1.PROMETHEUS_OPTIONS,
                optional: true,
            },
            ...extraInject,
        ],
    };
}
