// crystals-kyber-js/script/src/mlKem512Impl.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./errors.js", "./mlKem512Base.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.MlKem512Impl = void 0;
    exports.createMlKem512 = createMlKem512;
    const errors_js_1 = require("./errors.js");
    const mlKem512Base_js_1 = require("./mlKem512Base.js");
    /**
     * Synchronous implementation of MlKem512.
     *
     * Use {@link createMlKem512} to create an initialized instance.
     *
     * @example
     *
     * ```ts
     * // Using jsr:
     * import { createMlKem512 } from "@dajiaji/mlkem";
     * // Using npm:
     * // import { createMlKem512 } from "mlkem"; // or "crystals-kyber-js"
     *
     * const recipient = await createMlKem512();
     * const [pkR, skR] = recipient.generateKeyPair();
     *
     * const sender = await createMlKem512();
     * const [ct, ssS] = sender.encap(pkR);
     *
     * const ssR = recipient.decap(ct, skR);
     * // ssS === ssR
     * ```
     */
    class MlKem512Impl extends mlKem512Base_js_1.MlKem512Base {
        constructor() {
            super();
        }
        /** Generates a keypair [publicKey, privateKey]. */
        generateKeyPair() {
            try {
                return this._generateKeyPairCore();
            }
            catch (e) {
                throw new errors_js_1.MlKemError(e);
            }
        }
        /** Derives a keypair [publicKey, privateKey] deterministically from a 64-octet seed. */
        deriveKeyPair(seed) {
            try {
                return this._deriveKeyPairCore(seed);
            }
            catch (e) {
                throw new errors_js_1.MlKemError(e);
            }
        }
        /** Encapsulates: returns [ciphertext, sharedSecret]. */
        encap(pk, seed) {
            try {
                return this._encapCore(pk, seed);
            }
            catch (e) {
                throw new errors_js_1.MlKemError(e);
            }
        }
        /** Decapsulates: returns sharedSecret. */
        decap(ct, sk) {
            try {
                return this._decapCore(ct, sk);
            }
            catch (e) {
                throw new errors_js_1.MlKemError(e);
            }
        }
        /** @internal */
        static async _create() {
            const impl = new MlKem512Impl();
            await impl._setup();
            return impl;
        }
    }
    exports.MlKem512Impl = MlKem512Impl;
    /**
     * Creates a pre-initialized MlKem512 instance with synchronous operations.
     *
     * @returns A promise that resolves to an {@link MlKemInterface} instance.
     *
     * @example
     *
     * ```ts
     * const ctx = await createMlKem512();
     * const [pk, sk] = ctx.generateKeyPair();       // sync
     * const [ct, ssS] = ctx.encap(pk);              // sync
     * const ssR = ctx.decap(ct, sk);                // sync
     * ```
     */
    function createMlKem512() {
        return MlKem512Impl._create();
    }
});
