// crystals-kyber-js/script/mod.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "./src/errors.js", "./src/mlKem512.js", "./src/mlKem512Impl.js", "./src/mlKem768.js", "./src/mlKem768Impl.js", "./src/mlKem1024.js", "./src/mlKem1024Impl.js", "./src/sha3/sha3.js"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.shake256 = exports.shake128 = exports.sha3_512 = exports.sha3_256 = exports.createMlKem1024 = exports.MlKem1024 = exports.createMlKem768 = exports.MlKem768 = exports.createMlKem512 = exports.MlKem512 = exports.MlKemError = void 0;
    var errors_js_1 = require("./src/errors.js");
    Object.defineProperty(exports, "MlKemError", { enumerable: true, get: function () { return errors_js_1.MlKemError; } });
    var mlKem512_js_1 = require("./src/mlKem512.js");
    Object.defineProperty(exports, "MlKem512", { enumerable: true, get: function () { return mlKem512_js_1.MlKem512; } });
    var mlKem512Impl_js_1 = require("./src/mlKem512Impl.js");
    Object.defineProperty(exports, "createMlKem512", { enumerable: true, get: function () { return mlKem512Impl_js_1.createMlKem512; } });
    var mlKem768_js_1 = require("./src/mlKem768.js");
    Object.defineProperty(exports, "MlKem768", { enumerable: true, get: function () { return mlKem768_js_1.MlKem768; } });
    var mlKem768Impl_js_1 = require("./src/mlKem768Impl.js");
    Object.defineProperty(exports, "createMlKem768", { enumerable: true, get: function () { return mlKem768Impl_js_1.createMlKem768; } });
    var mlKem1024_js_1 = require("./src/mlKem1024.js");
    Object.defineProperty(exports, "MlKem1024", { enumerable: true, get: function () { return mlKem1024_js_1.MlKem1024; } });
    var mlKem1024Impl_js_1 = require("./src/mlKem1024Impl.js");
    Object.defineProperty(exports, "createMlKem1024", { enumerable: true, get: function () { return mlKem1024Impl_js_1.createMlKem1024; } });
    var sha3_js_1 = require("./src/sha3/sha3.js");
    Object.defineProperty(exports, "sha3_256", { enumerable: true, get: function () { return sha3_js_1.sha3_256; } });
    Object.defineProperty(exports, "sha3_512", { enumerable: true, get: function () { return sha3_js_1.sha3_512; } });
    Object.defineProperty(exports, "shake128", { enumerable: true, get: function () { return sha3_js_1.shake128; } });
    Object.defineProperty(exports, "shake256", { enumerable: true, get: function () { return sha3_js_1.shake256; } });
});
